ai_genomics/analysis/tag_evolution/tag_evolution_embeds.py

```diff
@@ -0,0 +1,244 @@
+"""This script clusters AI genomics entity embeddings using k-means at successive 
+timestamps and propagates the cluster labels across timestamps.
+
+It also saves out reduced entity embeddings and the propagated cluster labels to s3. 
+"""
+from ai_genomics.getters.data_getters import load_s3_data
+import pandas as pd
+from typing import Dict, List
+import itertools
+import numpy as np
+import pandas as pd
+import umap
+from sklearn.cluster import KMeans
+from sklearn.metrics import silhouette_score
+import ast
+
+from ai_genomics import bucket_name, logger
+from ai_genomics.getters.data_getters import load_s3_data, save_to_s3
+from ai_genomics.getters.patents import (
+    get_ai_genomics_patents,
+    get_ai_genomics_patents_entities,
+)
+from ai_genomics.getters.crunchbase import get_crunchbase_entities
+from ai_genomics.getters.gtr import get_gtr_entities
+from ai_genomics.getters.openalex import get_openalex_ai_genomics_works
+from ai_genomics.utils.bert_vectorizer import BertVectorizer
+from ai_genomics.utils.network_time_series import jaccard_similarity
+
+# instantiate models
+bert_model = BertVectorizer(multi_process=False).fit()
+reducer = umap.UMAP()
+
+# load entities
+patent_ents, crunchbase_ents, gtr_ents, oa_ents = (
+    get_ai_genomics_patents_entities(),
+    get_crunchbase_entities(),
+    get_gtr_entities(),
+    # temporarily as I wait for another PR to be merged
+    load_s3_data(
+        bucket_name, "outputs/entity_extraction/oa_ai_genomics_lookup_clean.json"
+    ),
+)
+
+
+def filter_data(
+    data: pd.DataFrame, query: str, date_col: str, id_col: str
+) -> pd.DataFrame:
+    """helper function to filter and rename columns across datasets"""
+    data = (
+        data.query(query)
+        .reset_index(drop=True)
+        .rename(columns={date_col: "date", id_col: "id"})
+    )
+    data["date"] = pd.to_datetime(data["date"]).dt.date
```

Comment:
this could be chained to the previous with `.assign(date = lambda df: pd.to_datetime(df["date"]).dt.date)`
]

and then you could just return the whole thing without assigning to `data` (which might by the way change the `data` df outside the function)